og_ui/src/Form/OgPermissionsForm.php

```diff
@@ -0,0 +1,337 @@
+          '#context' => [
+            'title' => $perm_item['title'],
+          ],
+        ];
+
+        // Show the permission description.
+        if (!$hide_descriptions) {
+          $form['permissions'][$perm]['description']['#context']['description'] = $perm_item['description'];
+          $form['permissions'][$perm]['description']['#context']['warning'] = $perm_item['warning'];
+        }
+
+        foreach ($roles as $rid => $role) {
+          list(,, $rid_simple) = explode('-', $rid, 3);
+
+          // The roles property indicates which roles the permission applies to.
+          $permission_applies = TRUE;
+          if (property_exists($permission, 'roles')) {
+            $target_roles = $permission->get('roles');
+            $permission_applies = empty($target_roles) || in_array($rid_simple, $target_roles);
+          }
+
+          if ($permission_applies) {
+            $form['permissions'][$perm][$rid] = [
+              '#title' => $role->getName() . ': ' . $perm_item['title'],
+              '#title_display' => 'invisible',
+              '#wrapper_attributes' => [
+                'class' => ['checkbox'],
+              ],
+              '#type' => 'checkbox',
+              '#default_value' => $role->hasPermission($perm) ? 1 : 0,
+              '#attributes' => ['class' => ['rid-' . $rid, 'js-rid-' . $rid]],
+              '#parents' => [$rid, $perm],
+            ];
+
+            // Show a column of disabled but checked checkboxes.
+            // Only applies to admins or default roles.
+            if ($roles[$rid]->get('is_admin') ||
+                in_array($rid_simple, $permission->getDefaultRoles())) {
+              $form['permissions'][$perm][$rid]['#disabled'] = TRUE;
+              $form['permissions'][$perm][$rid]['#default_value'] = TRUE;
+            }
+          }
+          else {
+            $form['permissions'][$perm][$rid] = [
+              '#title' => $role->getName() . ': ' . $perm_item['title'],
+              '#title_display' => 'invisible',
+              '#wrapper_attributes' => [
+                'class' => ['checkbox'],
+              ],
+              '#markup' => '-',
+            ];
+          }
+        }
+      }
+    }
+
+    $form['actions'] = ['#type' => 'actions'];
+    $form['actions']['submit'] = [
+      '#type' => 'submit',
+      '#value' => $this->t('Save permissions'),
+      '#button_type' => 'primary',
+    ];
+
+    $form['#attached']['library'][] = 'user/drupal.user.permissions';
+
+    return $form;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function submitForm(array &$form, FormStateInterface $form_state) {
+
+    /** @var \Drupal\og\Entity\OgRole $role */
+    foreach ($this->roles as $rid => $role) {
+      if (!$form_state->hasValue($rid)) {
+        continue;
```

Comment:
This is a bit concerning to me, it seems strange that the form state would not have a value for a particular role. I'm guessing this is done in case a role has been added inbetween the form was created and it was submitted?

Reply:
@zerolab, @pfrenssen, 

I think it was set for the case when a role was deleted so I suppose we can live with that.
If a role is added in between there will be no permissions for it, which we could also live with, I presume.

Or we could throw an exception here:

```suggestion
        $role_name = $role->getName();
        throw new OgRoleException("The role $role_name is missing.");
```

Note that even with this exception a race condition might (very unlikely but theoretically) still occur between loading and saving...